### PR TITLE
Metal: Add metal k8s resource provider

### PR DESCRIPTION
**Pull request by ecpullen**
**Issue number:**

#415 

**Description of changes:**

This adds the `metal-k8s-cluster-resource-agent` which takes and encoded management cluster kubeconfig, hardware csv, and eks anywhere config and creates a metal k8s cluster and registers all machines to ssm.

**Testing done:**

Created the test manifest with `cargo make create metal-sonobuoy-test`.
Applied the manifest with `cli run file <>`
Verified that the sonobuoy test passed, and the resource destruction was performed correctly.

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

**Comment by jpmcb** [comment]
Hmmm dependency mis/match? CI/CD failed with an openssh build failure:

> error: package `openssh v0.9.9` cannot be built because it requires rustc 1.63.0 or newer, while the currently active rustc version is 1.62.1

**Comment by ecpullen** [comment]
I added a pr https://github.com/bottlerocket-os/bottlerocket-test-system/pull/774 to update the rustc version to align with the Bottlerocket SDK.

File: bottlerocket/agents/Cargo.toml [hunk 1]
```diff
@@ -26,6 +26,7 @@ k8s-openapi = { version = "0.17", default-features = false, features = ["v1_24"]
 kube = { version = "0.78", default-features = false, features = ["config", "derive", "client"] }
 log = "0.4"
 maplit = "1"
+openssh = {version = "0.9.9", features = ["native-mux"]}
```
Comment (webern) [inline bottlerocket/agents/Cargo.toml]: nit:

Also does this introduce a requirement in the SDK?

```suggestion
openssh = { version = "0.9", features = ["native-mux"] }
```

File: bottlerocket/types/src/agent_config.rs [hunk 2]
```diff
@@ -182,6 +182,27 @@ pub struct VSphereK8sClusterConfig {
...
+    // Base64-encoded Kubeconfig for the CAPI management cluster
+    pub mgmt_cluster_kubeconfig_base64: String,
+
+    /// The role that should be assumed when activating SSM for the machines.
+    pub assume_role: Option<String>,
+
+    /// The base64-encoded EKS Anywhere config for this cluster.
+    pub cluster_config_base64: String,
+
+    /// The base64-encoded hardware csv that will be used for cluster creation.
```
Comment (webern) [inline bottlerocket/types/src/agent_config.rs:199]: This should probably be expanded on...

File: Dockerfile [hunk 3]
```diff
@@ -147,6 +147,35 @@ COPY --from=build-src /usr/share/licenses/testsys /licenses/testsys
...
 CMD dockerd --storage-driver vfs &>/dev/null & ./vsphere-k8s-cluster-resource-agent
 
+# =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^= =^..^=
+# Builds the Metal K8s cluster resource agent image
+FROM public.ecr.aws/amazonlinux/amazonlinux:2 as metal-k8s-cluster-resource-agent
+
+RUN yum install -y \
+        openssh-clients \
+        tar \
+        &&  yum clean all
```
Comment (jpculp) [inline Dockerfile]: super nit, but I would probably unindent this as it is not part of that first yum command. (also you have an extra space)

File: bottlerocket/samples/Makefile.toml [hunk 4]
```diff
@@ -133,6 +140,20 @@ if [ -n "${MGMT_CLUSTER_KUBECONFIG_PATH}" ]; then
 fi
 ''']}
 
+[tasks.hardware-csv-base64.env]
+HARDWARE_CSV_BASE64 = {script=['''
+if [ -n "${HARDWARE_CSV_PATH}" ]; then
+    cat ${HARDWARE_CSV_PATH} | base64 -w 0
```
Comment (jpculp) [inline bottlerocket/samples/Makefile.toml]: Not all versions of `base64` support the `-w` flag and instead have word wrapping by default. It's not the prettiest, but you might want to implement something similar to:
```
if base64 -w0 <<< "" &>/dev/null; then
    cat ${HARDWARE_CSV_PATH} | base64 -w0
else
    cat ${HARDWARE_CSV_PATH} | base64
fi
  ```

File: bottlerocket/samples/eks/metal-sonobuoy-test.yaml [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    keepRunning: true
+    privileged: true
+    configuration:
+      mgmtClusterKubeconfigBase64: ${MGMT_CLUSTER_KUBECONFIG_BASE64}
+      hardwareCsvBase64: ${HARDWARE_CSV_BASE64}
+      assumeRole: ${ASSUME_ROLE}
+      clusterConfigBase64: ${CLUSTER_CONFIG_BASE64}
+  dependsOn: []
+  destructionPolicy: onDeletion
+
```
Comment (jpculp) [inline bottlerocket/samples/eks/metal-sonobuoy-test.yaml]: Missing a newline.

File: bottlerocket/samples/Makefile.toml [hunk 6]
```diff
@@ -65,6 +71,7 @@ SONOBUOY_TEST_AGENT_IMAGE_URI = { value = "${AGENT_IMAGE_REGISTRY}/sonobuoy-test
 EKS_RESOURCE_AGENT_IMAGE_URI = { value = "${AGENT_IMAGE_REGISTRY}/eks-resource-agent:v${AGENT_IMAGE_VERSION}", condition = { env_not_set = ["EKS_RESOURCE_AGENT_IMAGE_URI"] } }
 VSPHERE_K8S_CLUSTER_RESOURCE_AGENT_IMAGE_URI = { value = "${AGENT_IMAGE_REGISTRY}/vsphere-k8s-cluster-resource-agent:v${AGENT_IMAGE_VERSION}", condition = { env_not_set = ["VSPHERE_K8S_CLUSTER_RESOURCE_AGENT_IMAGE_URI"] } }
 VSPHERE_VM_RESOURCE_AGENT_IMAGE_URI = { value = "${AGENT_IMAGE_REGISTRY}/vsphere-vm-resource-agent:v${AGENT_IMAGE_VERSION}", condition = { env_not_set = ["VSPHERE_VM_RESOURCE_AGENT_IMAGE_URI"] } }
+METAL_K8S_CLUSTER_RESOURCE_AGENT_IMAGE_URI = { value = "${AGENT_IMAGE_REGISTRY}/metal-k8s-cluster-resource-agent:v${AGENT_IMAGE_VERSION}", condition = { env_not_set = ["METAL_K8S_CLUSTER_RESOURCE_AGENT_IMAGE_URI"] } }
```
Comment (webern) [inline bottlerocket/samples/Makefile.toml:74]: I'm not asking for it to change now but I regret the formatting of this makefile. I generally like to spend currency in newlines to make things easier to read.

File: bottlerocket/samples/Makefile.toml (hunk 4 above)
Comment (jpculp) [reply bottlerocket/samples/Makefile.toml]: `echo "" | base64 -w0 >/dev/null 2>&1` might be even more compatible.

**Review by jpculp — APPROVED** [review]

**Review by webern — COMMENTED** [review]
This is great but we need documentation that describes what this is, how it works, that it uses EKS-A and CAPI, that it needs a pre-existing management cluster, etc.

File: bottlerocket/agents/Cargo.toml (hunk 1 above)
Comment (ecpullen) [reply bottlerocket/agents/Cargo.toml]: Not that I'm aware of.

File: bottlerocket/types/src/agent_config.rs (hunk 2 above)
Comment (ecpullen) [reply bottlerocket/types/src/agent_config.rs:199]:
```
/*!
This is a resource-agent for provisioning an EKS Anywhere Bare Metal k8s cluster.
This agent is used to provision workload clusters.
https://anywhere.eks.amazonaws.com/docs/concepts/clusterworkflow/
Before using this agent, a management EKS Anywhere cluster needs to be created. Its kubeconfig will
be passed into the agent with `encodedMgmtClusterKubeconfig`.
Once the management cluster has been provisioned, an EKS Anywhere config file needs to be created
for the cluster that needs to be created.
Additionally, a hardware csv corresponding to the cluster config needs to be base64 encoded.

This agent utilizes workload cluster creation by invoking `eksctl anywhere create cluster` with the
management cluster's kubeconfig, the cluster config, and the hardware csv.
After the cluster has been created, an AWS SSM activation is created for each provisioned node.
The ssh key provided by EKS Anywhere is used to ssh to the admin container and set the ssm activation
with `api client set` allowing ssm docs to be run on the Bare Metal machines.
In addition to the ssm activation, additional userdata can used with `userdata`.
The userdata is accepted in the form of base64 encoded toml and is converted from toml to json and
applied using `apiclient set --json` over ssh via the admin container.

At deletion, each SSM activation is terminated and the cluster is cleaned up using
`eksctl anywhere delete cluster`. There is a chance that some of the artifacts created by EKS Anywhere
are not cleaned up completely, so `kubectl delete -f` is used on the management cluster to ensure all
k8s artifacts included in the workload cluster config are deleted.
!*/
```

**Comment by ecpullen** [comment]
^ added a doc comment to the metal resource provider explaining prerequisites and how the agent works.

File: bottlerocket/agents/Cargo.toml (hunk 1 above)
Comment (jpculp) [reply bottlerocket/agents/Cargo.toml]: +1 to the version adjustment.

Comment (ecpullen) [reply bottlerocket/agents/Cargo.toml]: Missed it last time. Should be all set now.

**Review by jpculp — APPROVED** [review]

File: bottlerocket/types/src/agent_config.rs (hunk 2 above)
Comment (webern) [reply bottlerocket/types/src/agent_config.rs:199]: Looks good, thank you.

**Review by webern — APPROVED** [review]

File: bottlerocket/samples/Makefile.toml (hunk 6 above)
Comment (mjsterckx) [inline bottlerocket/samples/Makefile.toml:74]: This line will have to be copied into the `set-agent-images-single-repo` target after rebasing.

**Review by mjsterckx — COMMENTED** [review]
I think this should be rebased now that https://github.com/bottlerocket-os/bottlerocket-test-system/pull/772 is merged.

**Review by mjsterckx — APPROVED** [review]
LGTM

**Pull request merged**